src/main/java/org/dependencytrack/util/CacheStampedeBlocker.java

```diff
@@ -0,0 +1,260 @@
+                    cacheLoaders.put(key, cachePopulationFuture);
+                    rwLock.writeLock().unlock();
+                    locked = false;
+                    V result = retry.executeCallable(cacheLoader);
+                    cachePopulationFuture.getFuture().complete(result);
+                    Counter.builder("cache_stampede_blocker_load")
+                            .description("Total number of load event")
+                            .tags("cache", cacheName)
+                            .register(Metrics.getRegistry())
+                            .increment();
+                    // The completable future is not directly removed to account for the threads that detected that the cache
+                    // should be (re)loaded but will be scheduled a bit later. We don't want them to go through the locking-loading phase again.
+                    // Cleanup will be done by a dedicated thread
+                    return Optional.ofNullable(result);
+                }
+            } catch (Exception e) {
+                LOGGER.error("An error occurred while populating cache "+cacheName+" for key "+key+" : "+e.getMessage());
+                if (cachePopulationFuture != null && !cachePopulationFuture.future.isDone()) {
+                    cachePopulationFuture.getFuture().completeExceptionally(e);
+                    cacheLoaders.remove(key);
+                }
+                return Optional.empty();
+            } finally {
+                if(locked) {
+                    rwLock.writeLock().unlock();
+                }
+            }
+        }
+
+        if (cachePopulationFuture != null && blockCompetingThreads) {
+            LOGGER.debug("Cache "+cacheName+" population is ongoing or already finished for key "+key+" - Waiting patiently completion to return the value");
+            Counter.builder("cache_stampede_blocker_wait")
+                    .description("Total number of wait event")
+                    .tags("cache", cacheName)
+                    .register(Metrics.getRegistry())
+                    .increment();
+            try {
+                return Optional.ofNullable(cachePopulationFuture.getFuture().get());
+            } catch (InterruptedException| ExecutionException e) {
+                LOGGER.error("An error occurred while populating cache "+cacheName+" for key "+key+" : "+e.getMessage());
+                return Optional.empty();
+            }
+        } else {
+            LOGGER.debug("Cache "+cacheName+" population is ongoing or already finished for key "+key+" and stampede blocker configured to return immediately in this situation");
+            return Optional.empty();
+        }
+    }
+
+    public void cleanupExpirableFutureMap() {
+        LOGGER.debug("Starting to cleanup "+cacheName+"'s loader map");
+        List<K> keysToDelete =  cacheLoaders.entrySet().stream()
```

Comment:
I guess it could indeed happen that entries are removed while another thread is `put`ting entries into this map, or fetching data from it. Maybe `ConcurrentHashMap` would be a good choice?

Reply:
`cacheLoaders` is already a `ConcurrentHashMap` ;-)